### ISSUE TITLE
Prepare sections CSV for analysis

### ISSUE THREAD

**Issue by geohacker** [orma/openroads-vn-analytics#474]
From the ToR:

> Export geometries and attributes in CSV format periodically for feeding into analytics data pipeline

Next actions:
* [x] Build a script into tiler that queries all section attributes
* [x] Prepare a CSV and store on S3 with timestamp

**Comment by geohacker** [comment]
Hi @erikdao, I'll outline the CBA exports work in this ticket:
* The exports are part of the worker process that runs every 30 mins at the moment, but will be at a lower frequency of once per day. 
* The PR is https://github.com/orma/openroads-vn-tiler/pull/51.  
* The following is the header for [the CSV file](https://github.com/orma/openroads-vn-tiler/pull/51/files#diff-ef46868c79b20801f8a1b8aa6613d38eR12)
* I'll update the S3 location as soon as this is deployed. You should be able to do a basic http/s download.
* If you notice that this doesn't have enough data, that's purely because new datasets have not been imported yet. cc @PCDuc 

cc @olafveerman @kai2ser @batpad

**Comment by geohacker** [comment]
Just noticed a minor columns bug in the CSV, I'll fix this in a moment.

**Comment by geohacker** [comment]
Fixed the CSV errors. Sections export will start appearing everyday at https://s3-ap-southeast-1.amazonaws.com/openroads-vn-dumps-prod/cba/orma-sections-YYYY-MM-DD.csv

**Comment by geohacker** [comment]
This is done. Once @erikdao gets a chance to test, we'll reopen if there are things to address.

**Comment by olafveerman** [comment]
Feedback by DBC:
> I still have two question here:
> - Is there any tool in ORMA to make it or it would be done manually by the system administrator?
> - Is there any connection between ORMA and CBA module? (by connection I don't mean DB or API, I just mean the link of CBA in ORMA and vice-versa, or whether or not there is a scenarios to run CBA from ORMA)

cc @geohacker

**Comment by geohacker** [comment]
> Is there any tool in ORMA to make it or it would be done manually by the system administrator?

CBA section export is automatically generated everyday and stored on S3, accessible by the CBA program. 

> Is there any connection between ORMA and CBA module? (by connection I don't mean DB or API, I just mean the link of CBA in ORMA and vice-versa, or whether or not there is a scenarios to run CBA from ORMA)

My understanding is that CBA will directly use the sections exported by ORMA. We also hope to link to CBA from ORMA as a menu item when it is ready. Any other details will have to be confirmed by Coung.

**Comment by erikdao** [comment]
@geohacker Hi, The S3 CSV link seems not working for me.

**Comment by geohacker** [comment]
Hey @erikdao - looks like the cba script stopped working after a minor bug on November 20 but you can access files until then. For instance https://s3-ap-southeast-1.amazonaws.com/openroads-vn-dumps-prod/cba/orma-sections-2018-11-20.csv

I'm patching a fix now and this should be live later today. Thank you reopening the ticket!

**Comment by erikdao** [comment]
@geohacker Thanks for providing the sample file.

I've looked at the file and I can see that there are many missing values for the following parameters. Without those values, CBA cannot work.
* Number of Lanes Class
* Carriageway With
* Road Class
* Terrain Type
* Temperature Class
* Moisture Class
* Pavement Type
* Surface Type
* Pavement Condition Class
* Roughness (IRI)
* Traffic Level
* Traffic Annual Growth Scenario

And, with the following attributes, the values if presented would also be helpful
* Motor Cycle (veh/day)
* Small Car (veh/day)
* Medium Car (veh/day)
* Delivery Vehicle (veh/day)
* 4 Wheel Drive (veh/day)
* Light Truck (veh/day)
* Medium Truck (veh/day)
* Heavy Truck
* Articulated Truck
* Small Bus
* Medium Bus
* Large Bus

Another question for this data dump is that: Does this snapshot contains all sections data, or it just contains data uploaded for a certain day? And is there any available API for getting all the road sections of a particular province?

Thanks,

**Comment by geohacker** [comment]
> I've looked at the file and I can see that there are many missing values for the following parameters. Without those values, CBA cannot work.

@erikdao this is because the Vietnam team has not added any section level data yet to the database. This will be the mapping effort they have to do. Unfortunately, I do not have any insights on the timeline for this. I hope Kai or Duc can give you an idea.

> Another question for this data dump is that: Does this snapshot contains all sections data

The dump contains all data regardless of what's added that day or the previous day. This keeps it comprehensive. The files on S3 will then become a historical archive if anyone wants that. 

> API for getting all the road sections of a particular province?

The road sections per province are also stored and updated daily on S3. The URL format is https://openroads-vn-dumps-prod.s3.amazonaws.com/by-province-id/<province_id>.csv for example Thanh Hoa is https://openroads-vn-dumps-prod.s3.amazonaws.com/by-province-id/266.csv You can get all province ids from this endpoint http://api.openroads-vn.com/admin/stats

**Comment by erikdao** [comment]
@geohacker I'm bringing up this thread again. It seems that the archives after Nov 20 are still not available.
Can you check again and provide me with a full dump of all road assets currently on ORMA in CBA format?
Thanks,

**Comment by geohacker** [comment]
@erikdao sorry about that. I just noticed another hiccup. It is running now and the export process will take a little while before the CSV lands on S3. I'll keep you posted. Thank you.

**Comment by erikdao** [comment]
Thanks @geohacker for your quick response. Could we have this batch job running stable? At least until the end of this year for the demonstration purpose?

**Comment by geohacker** [comment]
Yes, I think once we resolve the ssl build issue today this will run stable, everyday.

**Comment by geohacker** [comment]
@erikdao thank you for your patience! The current process took a while to run but you have CSV over at https://s3-ap-southeast-1.amazonaws.com/openroads-vn-dumps-prod/cba/orma-sections-2018-12-18.csv - note that we're aware of the extraneous columns and are looking at it. But at the moment, the priority is to fix https://github.com/orma/openroads-vn-tiler/pull/70

**Comment by erikdao** [comment]
Hi @geohacker, the archive link seems to work again. I can download the archive for 2018-12-18.
However, when I take a further look at the CSV line, it seems that the first data row isn't separated from the header row as shown in the below screenshot. Could you double-check the file again?
![screen shot 2018-12-19 at 10 45 58](https://user-images.githubusercontent.com/26970731/50197842-893f5280-037b-11e9-9e84-b5093449c023.png)

**Comment by geohacker** [comment]
> Could you double-check the file again?

Yes I noticed this and we rolled out a fix as mentioned above. It's a silly one, but we're trying to get a build issue resolved before pushing this to production. You can see that here https://github.com/orma/openroads-vn-tiler/pull/70

Can you work with the current CSV for now? We'll have this resolved later today! Thanks!

**Comment by erikdao** [comment]
@geohacker Thanks for your prompt response. I just want to make sure that we have the correct CSV format to work on CBA. Actually, I can't work with these files now because, as mentioned above, the data is still missing many required attributes. Hope that the VN team could fill them in soon.
Thanks,

**Comment by erikdao** [comment]
Hi @geohacker,
I want to reiterate on this issue. As far as I can see the data dump is working properly, I can retrieve the data up to 2019-02-14. ;)
However, I still see that many required section attributes as mentioned previously are still missing. Is it because the mapping team hasn't updated the data?
Thanks,

**Comment by geohacker** [comment]
@erikdao thank you. You're right - the local team is working on adding more data but is currently paused because [bugs we are resolving](https://github.com/orma/openroads-vn-analytics/issues/545). These will be resolved next week, and you can expect more data hopefully. 

Also note that the data uploads are out of my control :)

**Comment by erikdao** [comment]
@geohacker Yeah. I understand that. I just wanted to make sure that there is no problem with ORMA regarding the data. Duc once told me that he uploaded many roads to the system and that the attributes should be presented.
Anyway, I'll check with them.
Thanks,
